### PR TITLE
[Feature][Connector-v2][PulsarSink]Add Pulsar Sink Connector.

### PR DESCRIPTION
**Pull request by lightzhao**
## Purpose of this pull request

Add Pulsar Sink Connector. #3018 
## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by lightzhao** [comment]
Hello everyone, ci check cannot automatically, ci check require approval?

**Comment by lightzhao** [comment]
Hello everyone, it has been running stably in the test environment for two weeks, who has time to help review.

**Review by hailin0 — COMMENTED** [review]
please update `release-note.md`
please add e2e testcase

reference
https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e

File: docs/en/connector-v2/sink/pulsar.md
Comment (hailin0) [inline docs/en/connector-v2/sink/pulsar.md]:
```suggestion
```


https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/concept/connector-v2-features.md#sink-connector-features

File: docs/en/connector-v2/sink/pulsar.md
Comment (hailin0) [inline docs/en/connector-v2/sink/pulsar.md]:
```suggestion
```

**Comment by lightzhao** [comment]
> please update `release-note.md` please add e2e testcase
> 
> reference https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e

waiting #3990 ,In this pr I added the pulsar-e2e module. doc has been updated.

**Comment by lightzhao** [comment]
Hello everyone, why can't the automatic ci check be performed after updating the code now, and it needs to be approved?

**Comment by TyrantLucifer** [comment]
> Hello everyone, why can't the automatic ci check be performed after updating the code now, and it needs to be approved?

Yes, infra changed the rule of this. It needs to be approved.

**Comment by lightzhao** [comment]
> > Hello everyone, why can't the automatic ci check be performed after updating the code now, and it needs to be approved?
> 
> Yes, infra changed the rule of this. It needs to be approved.

please review #3990，and then add pulsar sink e2e test. thanks.

**Comment by lightzhao** [comment]
please approve ci.

**Comment by lightzhao** [comment]
@Hisoka-X @hailin0 @TyrantLucifer please approve ci.

**Comment by lightzhao** [comment]
@Hisoka-X @hailin0 @EricJoy2048 please approve ci.

**Comment by lightzhao** [comment]
@TyrantLucifer @hailin0 @EricJoy2048 @Hisoka-X PTAL.

**Comment by lightzhao** [comment]
@all Can anyone help to review this pr?

**Comment by lightzhao** [comment]
@TyrantLucifer @hailin0 @EricJoy2048 @Hisoka-X PTAL.

**Comment by lightzhao** [comment]
It's been a long time, @TyrantLucifer @hailin0 @EricJoy2048 @Hisoka-X PTAL.

**Comment by ic4y** [comment]
Waiting for CI/CD.

**Comment by lightzhao** [comment]
@ic4y please approve ci. thanks.

**Comment by lightzhao** [comment]
@ALL  please approve ci. thanks.

**Comment by lightzhao** [comment]
image
Why is there such an exception? It has nothing to do with the module I submitted, and the latest code has also been merged.

**Comment by EricJoy2048** [comment]
> image Why is there such an exception? It has nothing to do with the module I submitted, and the latest code has also been merged.

Hi, @lightzhao  Please rebase dev branch and push again. @ic4y  Can you take a look about the CI error?

**Comment by lightzhao** [comment]
> > image Why is there such an exception? It has nothing to do with the module I submitted, and the latest code has also been merged.
> 
> Hi, @lightzhao Please rebase dev branch and push again. @ic4y Can you take a look about the CI error?

done. please ci.

**Comment by lightzhao** [comment]
@hailin0 @EricJoy2048 @Hisoka-X @ic4y PTAL.

**Comment by lightzhao** [comment]
@hailin0 @EricJoy2048 @Hisoka-X @ic4y PTAL.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/java/org/apache/seatunnel/e2e/connector/pulsar/PulsarSinkIT.java
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/java/org/apache/seatunnel/e2e/connector/pulsar/PulsarSinkIT.java:125]: consume data and check count & format

File: docs/en/connector-v2/sink/pulsar.md
Comment (hailin0) [inline docs/en/connector-v2/sink/pulsar.md]: update doc format 

https://github.com/apache/seatunnel/blob/dev/docs/en/connector-v2/sink/Kafka.md

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/java/org/apache/seatunnel/e2e/connector/pulsar/PulsarSinkIT.java
Comment (lightzhao) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/java/org/apache/seatunnel/e2e/connector/pulsar/PulsarSinkIT.java:125]: done

File: docs/en/connector-v2/sink/pulsar.md
Comment (lightzhao) [reply docs/en/connector-v2/sink/pulsar.md]: done

**Comment by lightzhao** [comment]
@hailin0 @EricJoy2048 @Hisoka-X @ic4y PTAL.

File: docs/en/connector-v2/sink/pulsar.md
Comment (Carl-Zhou-CN) [inline docs/en/connector-v2/sink/pulsar.md]: docs/en/connector-v2/sink/puslar.md -> Please refer to the "kafka.md" file for some upgrade instructions,
https://github.com/apache/seatunnel/blob/dev/docs/en/connector-v2/sink/Kafka.md

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkFactory.java
Comment (Carl-Zhou-CN) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkFactory.java:39]: Support TableSourceFactory/TableSinkFactory on connector，please refer to the following link for the upgrade instructions: https://github.com/apache/seatunnel/issues/5651.

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSink.java
Comment (Carl-Zhou-CN) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSink.java:46]: Support multi-table sink feature: please refer to the following link for the upgrade instructions: https://github.com/apache/seatunnel/issues/5652

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/state/PulsarSinkState.java
Comment (Carl-Zhou-CN) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/state/PulsarSinkState.java:32]: Wouldn't it be better to store the 'topic' as part of the state?

File: docs/en/connector-v2/sink/pulsar.md
Comment (lightzhao) [reply docs/en/connector-v2/sink/pulsar.md]: done.

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkFactory.java
Comment (lightzhao) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkFactory.java:39]: done.

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/state/PulsarSinkState.java
Comment (lightzhao) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/state/PulsarSinkState.java:32]: Only TxnID is required when commit, and the topic parameter is not required.

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSink.java
Comment (lightzhao) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSink.java:46]: Do kafka and pulsar need the support of multiple table sinks? Multiple topics are supported.

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkWriter.java
Comment (Carl-Zhou-CN) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkWriter.java:263]: The comment of these parameters is useless, can we delete it？

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkWriter.java
Comment (Carl-Zhou-CN) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkWriter.java]: The comment of these parameters is useless, can we delete it？

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkWriter.java
Comment (Carl-Zhou-CN) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkWriter.java]: Same as above

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkWriter.java
Comment (Carl-Zhou-CN) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkWriter.java]: Same as above

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/state/PulsarSinkState.java
Comment (Carl-Zhou-CN) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/state/PulsarSinkState.java:32]: ok

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSink.java
Comment (Carl-Zhou-CN) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSink.java:46]: @Hisoka-X cc

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkWriter.java
Comment (lightzhao) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSinkWriter.java]: done.

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSink.java
Comment (Hisoka-X) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/sink/PulsarSink.java:46]: Supporting multiple tables is mainly to accept multiple upstream tables, which can be implemented separately in the future.

**Comment by lightzhao** [comment]
@hailin0 @Carl-Zhou-CN @Hisoka-X PTAL.

**Review by Carl-Zhou-CN — DISMISSED** [review]
LGTM

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/resources/fake_to_pulsar.conf
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/resources/fake_to_pulsar.conf]:
```suggestion
 parallelism = 1
```

https://github.com/apache/seatunnel/pull/6003

**Review by hailin0 — COMMENTED** [review]
LGTM

**Review by EricJoy2048 — APPROVED** [review]

**Comment by lightzhao** [comment]
@hailin0 @Hisoka-X @Carl-Zhou-CN PTAL.

**Review by hailin0 — APPROVED** [review]

**Pull request merged**
